go/libraries/doltcore/sqle/database_provider.go

```diff
@@ -676,20 +612,101 @@ func (p *DoltDatabaseProvider) DropDatabase(ctx *sql.Context, name string) error
 		p.DropDatabaseHook(name)
 	}
 
-	// We not only have to delete this database, but any derivative ones that we've stored as a result of USE or
-	// connection strings
+	// We not only have to delete tracking metadata for this database, but also for any derivative
+	// ones we've stored as a result of USE or connection strings
 	derivativeNamePrefix := strings.ToLower(dbKey + dsess.DbRevisionDelimiter)
 	for dbName := range p.databases {
 		if strings.HasPrefix(strings.ToLower(dbName), derivativeNamePrefix) {
 			delete(p.databases, dbName)
 		}
 	}
-
 	delete(p.databases, dbKey)
 
 	return p.invalidateDbStateInAllSessions(ctx, name)
 }
 
+func (p *DoltDatabaseProvider) ListDroppedDatabases(ctx *sql.Context) ([]string, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	return p.droppedDatabaseManager.ListDroppedDatabases(ctx)
+}
+
+func (p *DoltDatabaseProvider) UndropDatabase(ctx *sql.Context, name string) (err error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newFs, exactCaseName, err := p.droppedDatabaseManager.UndropDatabase(ctx, name)
+	if err != nil {
+		return err
+	}
+
+	newEnv := env.Load(ctx, env.GetCurrentUserHomeDir, newFs, p.dbFactoryUrl, "TODO")
+	return p.registerNewDatabase(ctx, exactCaseName, newEnv)
+}
+
+// PurgeDroppedDatabases permanently deletes all dropped databases that have been stashed away in case they need
+// to be restored. Use caution with this operation – it is not reversible!
+func (p *DoltDatabaseProvider) PurgeDroppedDatabases(ctx *sql.Context) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	return p.droppedDatabaseManager.PurgeAllDroppedDatabases(ctx)
+}
+
+// registerNewDatabase registers the specified DoltEnv, |newEnv|, as a new database named |name|. This
+// function is responsible for instantiating the new Database instance and updating the tracking metadata
+// in this provider. If any problems are encountered while registering the new database, an error is returned.
+func (p *DoltDatabaseProvider) registerNewDatabase(ctx *sql.Context, name string, newEnv *env.DoltEnv) (err error) {
+	// This method MUST be called with the provider's mutex locked. TryLock allows us to validate that the
+	// mutex is locked (without actually locking it and causing a deadlock) and to error out if we detect
+	// that the mutex is NOT locked.
+	if p.mu.TryLock() {
+		defer p.mu.Unlock()
```

Comment:
This seems like an anti-pattern. Fine to check if the lock is held, but if we didn't take the lock ourselves we probably shouldn't be releasing it either. Up to the caller

Reply:
This is probably worth a comment since it's not a common pattern... `TryLock` will attempt to lock the mutex and if it is able to, then it executes this block. That means we need to error out, but we've also grabbed the mutex in the process of calling `TryLock` to test the mutex, so this function needs to release the lock and not just leave it locked. Pulling this out into a helper function would probably make the coder a little easier to read and avoid needing to explain that behavior in multiple places. 

I'm not committed this to this approach either, but it seems like doing something to ensure the lock is held is worth it.

Reply:
Cleaned this up in a47c4f8b7c656d302b307f9c3cce1413bc51a929. That removed some duplication, will make that util function easier to reuse in the future, and the comments should better explain why the lock needs to be released in that function (since it was grabbed in that function).